test/custom-task-ctrls/wait-task/pkg/reconciler/reconciler_test.go

```diff
@@ -260,3 +232,150 @@ status:
 		})
 	}
 }
+
+func TestReconcile_Retries(t *testing.T) {
+	for _, tc := range []struct {
+		name                string
+		duration            string
+		timeout             string
+		retries             int
+		params              string
+		wantStatus          string
+		isParentPRCancelled bool
+	}{{
+		name:     "retry when timeout",
+		duration: "2s",
+		timeout:  "1s",
+		retries:  1,
+		wantStatus: fmt.Sprintf(`
+status:
+  conditions:
+  - reason: %s
+    status: %q
+    type: %s
+  observedGeneration: 0
+  retriesStatus:
+  - conditions:
+    - reason: %s
+      status: %q
+      type: %s
+`, "TimedOut", corev1.ConditionFalse, apis.ConditionSucceeded, "TimedOut", corev1.ConditionFalse, apis.ConditionSucceeded),
+		isParentPRCancelled: false,
+	}, {
+		name:     "don't retry if retries unspecified",
+		duration: "2s",
+		timeout:  "1s",
+		wantStatus: fmt.Sprintf(`
+status:
+  conditions:
+  - reason: %s
+    status: %q
+    type: %s
+  observedGeneration: 0
+`, "TimedOut", corev1.ConditionFalse, apis.ConditionSucceeded),
+		isParentPRCancelled: false,
+	}, {
+		name:     "don't retry when canceled",
+		duration: "2s",
+		wantStatus: fmt.Sprintf(`
+status:
+  conditions:
+  - reason: %s
+    status: %q
+    type: %s
+  observedGeneration: 0
+`, "Cancelled", corev1.ConditionFalse, apis.ConditionSucceeded),
+		isParentPRCancelled: true,
+	}} {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			rec := &Reconciler{
+				Clock: testClock,
+			}
+
+			runName := helpers.ObjectNameForTest(t)
+			runYAML := fmt.Sprintf(`
+metadata:
+  name: %s
+spec:
+  retries: %d
+  timeout: %s
+  ref:
+    apiVersion: %s
+    kind: %s
+  params:
+  - name: duration
+    value: %s
+`, runName, tc.retries, tc.timeout, apiVersion, kind, tc.duration)
+			r := parse.MustParseRun(t, runYAML)
+			if tc.isParentPRCancelled {
+				r.Spec.Status = v1alpha1.RunSpecStatusCancelled
+				r.Spec.StatusMessage = v1alpha1.RunCancelledByPipelineMsg
+			}
+
+			fakeReconcile(t, ctx, rec, r)
+
+			// Compose expected Run
+			wantRunYAML := fmt.Sprintf(`
+metadata:
+  name: %s
+spec:
+  retries: %d
+  timeout: %s
+  ref:
+    apiVersion: %s
+    kind: %s
+  params:
+  - name: duration
+    value: %s
+`, runName, tc.retries, tc.timeout, apiVersion, kind, tc.duration)
+			wantRunYAML = wantRunYAML + tc.wantStatus
+			wantRun := parse.MustParseRun(t, wantRunYAML)
+			if tc.isParentPRCancelled {
+				wantRun.Spec.Status = v1alpha1.RunSpecStatusCancelled
+				wantRun.Spec.StatusMessage = v1alpha1.RunCancelledByPipelineMsg
+			}
+
+			if d := cmp.Diff(wantRun, r,
+				filterTypeMeta,
+				filterObjectMeta,
+				filterCondition,
+				filterRunStatus,
+			); d != "" {
+				t.Errorf("-got +want: %v", d)
+			}
+		})
+	}
+}
+
+func fakeReconcile(t *testing.T, ctx context.Context, rec *Reconciler, r *v1alpha1.Run) {
```

Comment:
I probably missed this on a different code review but why is this function necessary? it looks like it's testing the timeout behavior, but you can set the start time on the run to whatever you want (e.g. 10 minutes ago) when calling Reconcile. I'm concerned that moving some of the "reconcile" behavior to the test cases can hide issues-- dependency injection is probably a better pattern.

Reply:
Thanks for the comment Lee! Actually I was not trying to test the timeout function, I tried to simulate requeue events triggered by controller.EnqueueAfter as well as controller.EnqueueImmidiately, without modifying anything of the Run in this func. But it does duplicate some logics in the reonciler -- the timing of requeue the event which can lead to some potential issues in the future though.

I feel like it might not proper to update it in this PR because it requires to update all of the existing test cases, while this PR only focuses on the `retries`. I can open another PR to update it. wdyt?

Reply:
ah ok makes sense-- I think it would be better to write a test case that reconciles an already-over-timeout run and verify that the reconciler marks it as timed out. I agree this belongs in a separate PR.

Reply:
SG!

Reply:
would you mind opening a PR to fix the existing test cases before merging this PR? (If that's too much, you can open a tracking issue instead and link it here)

Reply:
SG, created an issue https://github.com/tektoncd/pipeline/issues/5456 to track it.

Reply:
Done